dbms/src/Interpreters/Join.cpp

```diff
@@ -251,8 +280,39 @@ void Join::setSampleBlock(const Block & block)
             key_columns[i] = &static_cast<const ColumnNullable &>(*key_columns[i]).getNestedColumn();
     }
 
-    /// Choose data structure to use for JOIN.
-    init(chooseMethod(key_columns, key_sizes));
+    if (strictness == ASTTableJoin::Strictness::Asof)
+    {
+        if (kind != ASTTableJoin::Kind::Left)
```

Comment:
I agree that there is value in supporting OUTER and INNER (for completeness and increased performance later down the line).

- OUTER would insert nulls on both left and right side if it cannot find a value.
  This would require to also maintain a BST of the entire left hand side, which (1) is quite expensive and (2) would need some code duplication or if constexpr to make work.
- INNER would not have nulls either left or right if it cannot find any values (and could be implemented as a WHERE clause in the query on top of a LEFT or OUTER join)
- LEFT (as it is currently) can only have nulls on the right hand side.

I don't follow your logic why you would want to force OUTER ASOF to be LEFT and normal ASOF to be a special INNER. From a users perspective it seems you would want the minimal deviation from the result (eg. for OUTER you can have nulls on both sides, for INNER you cannot have any and for LEFT you can only have nulls on the right hand side).

From my perspective as user, the ASOF LEFT join is the only thing really needed. You pretty much always want to select either the left or right side as the resulting values in time for your series (which makes OUTER obsolete in most cases and RIGHT in all cases, as you can reverse left and right). Furthermore, INNER is essentially a LEFT join followed by doing a WHERE right != null (though for performance you might want to implement INNER directly)

Reply:
OK, but there's two reasons to make INNER. First, as you have said, there could be a perfomance penalties for nullable columns. Second, if I interested in graphics it's better not to set some point then set useless nulls or defaults.

Reply:
I'll try to add INNER join then - it shouldn't be much work (hopefully) as the Join code is already very nice to read and modular (props for that). OUTER would be quite a bit more work so I'll leave that for some later time as it is needed.

Reply:
Done e7a10b8a3
Turns out it was already working, just had to remove the check against INNER joins.